src/vsr/replica.zig

```diff
@@ -7659,6 +7610,92 @@ pub fn ReplicaType(
+            //
+            // DVC headers are stitched together from the journal and existing view headers (they
+            // might belong to the next log wrap), to guarantee that a DVC with log_view=v includes
+            // all uncommitted ops with views <v. Special case: a primary that has collected a DVC
+            // quorum and installed surviving headers into the journal ignores view headers (they
+            // might have some truncated ops).
+            var view_headers_updated = vsr.Headers.ViewChangeArray{
+                .command = .do_view_change,
+                .array = .{},
+            };
+
+            const view_headers_op_max = self.view_headers.array.get(0).op;
+            var op = if (primary_repairing) self.op else @max(self.op, view_headers_op_max);
+            for (0..constants.pipeline_prepare_queue_max + 1) |_| {
+                const header_journal: ?*const vsr.Header.Prepare =
+                    self.journal.header_with_op(op);
+
+                const header_view: ?*const vsr.Header.Prepare = header: {
+                    if (primary_repairing or op > view_headers_op_max) break :header null;
+
+                    const header = &self.view_headers.array.const_slice()[view_headers_op_max - op];
+                    break :header switch (vsr.Headers.dvc_header_type(header)) {
+                        .valid => header,
+                        .blank => null,
+                    };
+                };
+
+                if (header_journal != null and header_view != null) {
+                    assert(header_journal.?.op == header_view.?.op);
+                    assert(header_journal.?.view == header_view.?.view);
+                    assert(header_journal.?.checksum == header_view.?.checksum);
+                }
+
+                if (header_journal == null and header_view == null) {
+                    view_headers_updated.append_blank(op);
+                } else {
+                    if (header_journal) |h| {
+                        view_headers_updated.append(h);
+                    } else {
+                        // Transition from normal status, but the SV headers were part of the next
+                        // wrap, so we didn't install them to our journal, and we didn't catch up.
+                        // We will reuse the SV headers as our DVC headers to ensure that
+                        // participating in another view-change won't allow the op to backtrack.
+                        assert(self.log_view == self.view);
+                        view_headers_updated.append(header_view.?);
+                    }
+                }
+
+                if (op <= self.commit_max) break;
+                op -= 1;
+            } else unreachable;
+
+            assert(op <= self.commit_max);
+            assert(op == self.commit_max or self.commit_max > self.op);
+
+            self.view_headers = view_headers_updated;
+            self.view_headers.verify();
```

Comment:
Alternatively, this function could just return `view_headers_updated` and let `transition_to_view_change_status` perform the actual update, so that it is more clear at the call site that that is what is happening.

Reply:
I'd rather keep the signature consistent with `primary_update_view_headers`.